src/mlpack/core/optimizers/katyusha/katyusha_impl.hpp

```diff
@@ -0,0 +1,188 @@
+    tolerance(tolerance),
+    shuffle(shuffle)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename DecomposableFunctionType>
+double Katyusha::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+  const double tau1 = 0.5 - tau;
+  const double tau2 = 1 + lambda * stepSize;
+
+  // To keep track of where we are and how things are going.
+  size_t currentFunction = 0;
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Calculate the first objective function.
+  for (size_t i = 0; i < numFunctions; i += batchSize)
+  {
+    const size_t effectiveBatchSize = std::min(batchSize, numFunctions - i);
+    overallObjective += function.Evaluate(iterate, i, effectiveBatchSize);
+  }
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat fullGradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+  arma::mat iterate0;
+
+  arma::mat y = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+  arma::mat z = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+  arma::mat w = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; /* incrementing done manually */)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
+    {
+      // Output current objective function.
+      Log::Info << "Katyusha: iteration " << i << ", objective "
+          << overallObjective << "." << std::endl;
+
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "Katyusha: converged to " << overallObjective
+            << "; terminating  with failure.  Try a smaller step size?"
+            << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "Katyusha: minimized within tolerance " << tolerance
+            << "; terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentFunction = 0;
+
+      if (shuffle) // Determine order of visitation.
+        function.Shuffle();
+    }
+
+    iterate0 = iterate;
+    y = iterate;
+    z = iterate;
+    w.zeros();
+
+    // Compute the full gradient.
+    size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+    function.Gradient(iterate, 0, fullGradient, effectiveBatchSize);
```

Comment:
The idea is to avoid an extra copy for the accumulation inside the loop:

```c++
function.Gradient(iterate, f, gradient, effectiveBatchSize);
fullGradient += gradient;
```

Maybe I missed something?

Reply:
Oh!  Now I understand.  Thanks for clarifying. :)